src/modules/Dashboard/features/Applet/Activities/Activities.tsx

```diff
@@ -1,59 +1,93 @@
 import { useEffect, useMemo, useState } from 'react';
+import { useTranslation } from 'react-i18next';
 import { useParams } from 'react-router-dom';
 
-import { getAppletActivitiesApi } from 'api';
+import { getAppletActivitiesApi, getAppletApi } from 'api';
+import { Spinner } from 'shared/components';
 import { useAsync } from 'shared/hooks';
 import {
   ActivitiesData,
   ActivityGrid,
   useActivityGrid,
 } from 'modules/Dashboard/components/ActivityGrid';
+import { FlowGrid } from 'modules/Dashboard/components/FlowGrid';
+import { Activity, ActivityFlow } from 'redux/modules';
+import { StyledFlexColumn } from 'shared/styles';
+
+import { ActivitiesSectionHeader } from './ActivitiesSectionHeader';
+import { ActivitiesToolbar } from './ActivitiesToolbar';
 
 export const Activities = () => {
   const { appletId } = useParams();
+  const { t } = useTranslation('app');
+  const [flows, setFlows] = useState<ActivityFlow[]>([]);
+  const [flowActivities, setFlowActivities] = useState<Activity[]>([]);
   const [activitiesData, setActivitiesData] = useState<ActivitiesData | null>(null);
-  const [isLoading, setIsLoading] = useState(true);
   const dataTestId = 'dashboard-applet-activities';
-
   const { formatRow, TakeNowModal } = useActivityGrid(dataTestId, activitiesData);
 
-  const { execute: getActivities } = useAsync(
+  const { execute: getActivities, isLoading: isLoadingActivities } = useAsync(
     getAppletActivitiesApi,
     (response) => {
       const activitiesDetails = response?.data.result.activitiesDetails;
 
       return setActivitiesData({ result: activitiesDetails, count: activitiesDetails.length });
     },
-    undefined,
-    () => setIsLoading(false),
   );
 
+  const { execute: getFlows, isLoading: isLoadingFlows } = useAsync(getAppletApi, (response) => {
+    if (response?.data?.result) {
+      setFlows(response.data.result.activityFlows ?? []);
+      setFlowActivities(response.data.result.activities ?? []);
+    }
+  });
+
+  const isLoading = isLoadingActivities || isLoadingFlows;
+
   useEffect(() => {
     if (!appletId) return;
-    getActivities({
-      params: {
-        appletId,
-      },
-    });
+    getFlows({ appletId });
+    getActivities({ params: { appletId } });
 
     return () => {
       setActivitiesData(null);
     };
-  }, [appletId, getActivities]);
+  }, [appletId, getActivities, getFlows]);
 
   const activities = useMemo(
     () => (activitiesData?.result ?? []).map((activity) => formatRow(activity)),
     [activitiesData, formatRow],
   );
 
   return (
-    <ActivityGrid
-      rows={activities}
-      TakeNowModal={TakeNowModal}
-      data-testid={dataTestId}
-      isLoading={isLoading}
-      order={'desc'}
-      orderBy={''}
-    />
+    <StyledFlexColumn sx={{ gap: 2.4, overflow: 'auto', p: 3.2 }}>
+      <ActivitiesToolbar appletId={appletId} data-testid={dataTestId} />
+
+      {isLoading ? (
+        <Spinner />
+      ) : (
```

Comment:
Nit: Just noting that we're either rendering the spinner, or the populated activities – however in many other screens, the spinner is just overlaid on top of the content (if any has been rendered yet).

As it is now (using either-or logic), if the activities are ever refreshed from the API after the page has first rendered (which won't happen here yet, but could later as we further build out this screen), then the existing rendered content would flicker away during refresh. NBD, but it might be nicer to avoid the flicker down the road.

Reply:
Hm, this was a bit more nuanced than it first appeared. I set up the loading states to mimic the existing first-time loading experience (eg, the toolbar and page UI appears with the spinner, and then the data renders when available), but you're right that I didn't account for the possibility of persisting data when re-fetching without leaving the page.

That said, I'm also not sure the existing implementation allowed for this, either — the `useAsync` hook we're using here [clears any existing data](https://github.com/ChildMindInstitute/mindlogger-admin/blob/60bb60127f708a6c8ea01ecade05dae91a1ae459/src/shared/hooks/useAsync.ts#L22) whenever `execute` is called.

It's also a bit more nuanced than it first appears on this page, because there's now multiple parallel network requests that will return at different times, and I'm trying to avoid some disruptive shifting UI that occurs when one resolves before the other or vice/versa.

I've handled it in this PR by adding an apollo-client/useQuery-inspired `previousData` property to the return value of `useAsync`, [here](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1691/commits/3660c2ad5b6693e558fe99041d2ca0cbe3982029), which I'm using both to persist data for subsequent requests, and infer nuanced loading state info (ie, if it is the first or subsequent time `execute` has been called) to render the UI as you're describing, here.